### ISSUE TITLE
Bug: Can't open SquareDesk, related to Wi-Fi

### ISSUE THREAD

**Issue by pkbott** [mpogue2/SquareDesk#1132]
This morning I added tags (years to all of my music already had some but did the rest) now Square Desk will not even open at all.  Then copied the square desk Music folder to my backup machine and now neither will open square desk.  Have a dance Friday and really need at least one to work.  HELP!

**Comment by pkbott** [comment]
![Uploading Screenshot 2024-06-17 at 10.55.03 AM.png…]()

**Comment by mpogue2** [comment]
Yes, please go back to previous version, until we can figure out what is going on!

**Comment by mpogue2** [comment]
If you see that message again, make sure that you are not running two copies at the same time.  Check for that, then Ok to click "continue", which should bring it up normally!

**Comment by pkbott** [comment]
Sorry to ask, but is it really as easy as just choosing a previous version and reinstalling? Or do I need to delete what I have currently after backing up my music library and then choose a version and then copy in my music library?What version was the most stable in your opinion since I came up a long ways?PaulReady to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 17, 2024, at 11:53 AM, Mike Pogue ***@***.***> wrote:﻿
Yes, please go back to previous version, until we can figure out what is going on!

**Comment by danlyke** [comment]
Yes it is really as easy as just choosing a version and reinstalling. It's
been a long time since we changed the database format, and even then we've
been pretty careful to allow backwards and forwards compatibility on
everything. If you have a previous .dmg in your downloads folder that
worked for you, that's probably a good bet!

**Comment by pkbott** [comment]
That’s the strange part. I wasn’t running two copies yet. It acted like I was the other interesting thing was my Wi-Fi was shut off as if it was running a second copy under force quit there was nothing there.  I was using the option to turn off the Wi-Fi in the preferences of Square desk if that leads you to some other bugs. Ready to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 17, 2024, at 11:55 AM, Mike Pogue ***@***.***> wrote:﻿
If you see that message again, make sure that you are not running two copies at the same time.  Check for that, then Ok to click "continue", which should bring it up normally!

**Comment by pkbott** [comment]
Thanks I usually don’t keep the DMG files. I will going forward but I do not have a current release. I’ll go onto get hub site and see if I can find something again a stable release. I’m assuming 1.7 possibly I had .9 I believe is what I was still running. Ready to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 17, 2024, at 12:25 PM, Paul ***@***.***> wrote:﻿That’s the strange part. I wasn’t running two copies yet. It acted like I was the other interesting thing was my Wi-Fi was shut off as if it was running a second copy under force quit there was nothing there.  I was using the option to turn off the Wi-Fi in the preferences of Square desk if that leads you to some other bugs. Ready to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 17, 2024, at 11:55 AM, Mike Pogue ***@***.***> wrote:﻿
If you see that message again, make sure that you are not running two copies at the same time.  Check for that, then Ok to click "continue", which should bring it up normally!

**Comment by pkbott** [comment]
Well I guess I did keep the last one, went back to 1.05 and all is ok again
it seems.

Paul

**Comment by mpogue2** [comment]
Thanks for letting us know....it's pretty weird what you're seeing (I have been running this version at all the dances I call for, for weeks, with zero errors...I must be lucky! :-)

Paul, could you kindly post the version of Mac OS and which laptop that you are on right now?  I am running Sonoma 14.5 on an M1 MacBook Air laptop.  I also ran the 1.0.8 SquareDesk in a virtual machine (also Sonoma 14.5), with no problems.  So, it looks like this is gonna be one of those hard-to-find "it only happens on one machine" problems...
Sorry that one machine is you! :-)

Yes, you can always go backward to a previous version.  I keep many versions on my laptop, but I only run one at a time.  And, if there's a bug, I can go backward to see if that bug was there in previous versions -- this helps me to isolate the bug to a single change that I made...

Thanks again for your help!!  More bug reports appreciated!

**Comment by pkbott** [comment]
Running same OS as you. I’m the lucky one though 🤣Ready to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 18, 2024, at 5:13 PM, Mike Pogue ***@***.***> wrote:﻿
Thanks for letting us know....it's pretty weird what you're seeing (I have been running this version at all the dances I call for, for weeks, with zero errors...I must be lucky! :-)
Paul, could you kindly post the version of Mac OS and which laptop that you are on right now?  I am running Sonoma 14.5 on an M1 MacBook Air laptop.  I also ran the 1.0.8 SquareDesk in a virtual machine (also Sonoma 14.5), with no problems.  So, it looks like this is gonna be one of those hard-to-find "it only happens on one machine" problems...
Sorry that one machine is you! :-)
Yes, you can always go backward to a previous version.  I keep many versions on my laptop, but I only run one at a time.  And, if there's a bug, I can go backward to see if that bug was there in previous versions -- this helps me to isolate the bug to a single change that I made...
Thanks again for your help!!  More bug reports appreciated!

**Comment by pkbott** [comment]
Sorry same OS as youUsing an M1 MacBook AirWondering if there’s a way I should clean out that entire program and restart?Ready to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 18, 2024, at 5:33 PM, Paul ***@***.***> wrote:﻿Running same OS as you. I’m the lucky one though 🤣Ready to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 18, 2024, at 5:13 PM, Mike Pogue ***@***.***> wrote:﻿
Thanks for letting us know....it's pretty weird what you're seeing (I have been running this version at all the dances I call for, for weeks, with zero errors...I must be lucky! :-)
Paul, could you kindly post the version of Mac OS and which laptop that you are on right now?  I am running Sonoma 14.5 on an M1 MacBook Air laptop.  I also ran the 1.0.8 SquareDesk in a virtual machine (also Sonoma 14.5), with no problems.  So, it looks like this is gonna be one of those hard-to-find "it only happens on one machine" problems...
Sorry that one machine is you! :-)
Yes, you can always go backward to a previous version.  I keep many versions on my laptop, but I only run one at a time.  And, if there's a bug, I can go backward to see if that bug was there in previous versions -- this helps me to isolate the bug to a single change that I made...
Thanks again for your help!!  More bug reports appreciated!

**Comment by mpogue2** [comment]
No need to clean out old program.  Multiple versions can be on the disk at the same time, it's just not a good idea to run more than one of them at the same time.

I'm thinking what might have happened is that SquareDesk crashed, but didn't completely quit, and when you started up a second copy, you had 2 running at the same time, which might have confused the second copy.  I'll investigate the crash, and given that you're running the same exact Mac OS version on the same M1 Air hardware, hopefully I can reproduce that problem here!

**Comment by mpogue2** [comment]
Also, if you get that "database is in use" message, just hit Continue, and it should be OK, as long as the other copy of SquareDesk is not actively doing anything to the database (and it probably isn't, if it's a zombie copy!).  This message is mainly for people who run SquareDesk on more than one machine, with a cloud storage service (e.g. Dropbox, Box, iCloud, etc.) hosting the music directory.  We don't want two machines both working on the database at the same time, with the cloud storage service moving data back and forth at the same time....

**Comment by pkbott** [comment]
I did get the message and I did hit continue and it would not continue. Don’t recall what was going on at the moment but it had definitely crashed when that happened so it’s possible that what you’re saying is true two copies were still running or not completely quit on one.Ready to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Jun 22, 2024, at 1:52 AM, Mike Pogue ***@***.***> wrote:﻿
Also, if you get that "database is in use" message, just hit Continue, and it should be OK, as long as the other copy of SquareDesk is not actively doing anything to the database (and it probably isn't, if it's a zombie copy!).  This message is mainly for people who run SquareDesk on more than one machine, with a cloud storage service (e.g. Dropbox, Box, iCloud, etc.) hosting the music directory.  We don't want two machines both working on the database at the same time, with the cloud storage service moving data back and forth at the same time....

**Comment by mpogue2** [comment]
I've done some work on trying to reproduce this.  I am suspecting that it was related to the "auto Wi-Fi disable" feature, which I don't use myself.  But, even with this feature ON, it worked every time for me.

@pkbott Are you also switching frequently between Light and Dark modes?  IIRC, you said you are doing that, which I also don't do, so there might be some interaction there between WiFi and switching Light/Dark modes.  I have been unable to reproduce this, but I don't doubt that you are seeing this problem!

**Comment by mpogue2** [comment]
Ah!  I just saw a hang, and here is the report when I did Force Quit.  It was definitely in Light Mode, and I was NOT switching from Light to Dark or vise versa.  I had not used it in a while, or maybe I quit it.  It looks like it might be stuck in destructors?  I've never seen this before in Dark Mode, maybe that's a clue?

[bug_1132.txt](https://github.com/user-attachments/files/16106577/bug_1132.txt)

**Comment by mpogue2** [comment]
**Another key point:**  when SquareDesk hangs on the way down (e.g. when switching Light <-> Dark modes), it might not release the lock file.  So, the next time we try to start it up, we get the "SquareDesk database is in use" message (because the hung copy of SquareDesk has not released the lock).  So, a hang is consistent with the symptoms so far.

I'm also seeing a error message in the log:

"warning: iofull was null ()"

which suggests that SD is hung.

If I am interpreting the hang report correctly, it's stuck inside ~SDThread, which looks like this:

```
    mutexSDAwaitingInput.unlock();
    mutexAckToMainThread.unlock();
    QMutexLocker locker(&mutexThreadRunning);
    if (!wait(250))
    {
        qWarning() << "Thread unable to stop, calling terminate";
        terminate();
    }
    numberOfSDThreadsActive--;
```

I am not quite sure why we're trying to take the existing lock on the way out.  If the thread is still running, because SD is hung, mutexThreadRunning does NOT get released, and I would think that ~SDThread would be stuck forever waiting to acquire the same lock.  Perhaps we should just delete this re-acquire, and let it wait() for 250ms, and then forcibly terminate the thread (note that documentation says that terminate() can be dangerous, because it's not a clean exit, but we're already in the non-clean-exit mode, right?)?

**I think that a race condition is a plausible explanation for the hang:**
- user says "switch to Light/Dark mode"
- main.cpp closes the mainwindow, and does startDetached.  
- BUT, the app isn't really shut down yet, because ~SDThread is hung waiting for sd to die (we are running a new copy of SD in this release, right?)
- so, the SD thread doesn't get killed, even though finishAndShutdownSD was called.  iofull error message is a hint that maybe our sending "exit" to SD isn't gonna work.
- clearLockFile doesn't clear the database lock, which causes the error message

**Possible things to do:**
- call clearLockFile as soon as we get into the destructor for MainWindow, to make it less likely that restarting will give us the scary error message
- don't reacquire the mutexThreadRunning, because we really don't care whether SD is running or not.  Instead, if the thread isRunning(), wait 250ms and then terminate() it anyway.  Otherwise, the thread is already stopped, and we can just exit the destructor without explicitly terminating.
- update to a newer version of SD? (39.66 is the latest as of May 2024)
- update to a newer version of Qt?  (Qt6.7.2 is now available as of June 2024, with a boatload of bug fixes)

**Comment by mpogue2** [comment]
Fixed by: 1747c06bc5837df84d8f2232888c7f5390b67d19

- Moved cleanLockFile up, so it would clear right away in the destructor.  
- We now do NOT try to acquire the mutexThreadRunning in the SDThread destructor, because that could hang if SD was itself hung, preventing shutdown.  
- Now we check every time iofull is dereferenced, because sometimes it could be null (if now if it was, it's logged in ALL places that might happen).  
- Added code to do a restart Light/Dark mode test (every 5 seconds, reboot and switch modes), and ran that test more than 200 times (now, no crashes, and it works every time, whereas before we were indeed seeing iofull be null sometimes).  Tested both with auto-Wi-Fi disabling ON and OFF (it crashed more often with it on, which is why I think @pkbott saw this problem and I did not).
